### ISSUE TITLE
Ignore ruby notation

### ISSUE THREAD

**Issue by nicolasmaia** [AwesomeTTS/awesometts-anki-addon#9]
When AwesomeTTS sees text like this: 
![captura de tela de 2017-06-23 13-25-50](https://user-images.githubusercontent.com/23157217/27491145-b3c03a30-5830-11e7-888a-eb26bb2f2521.png)
HTML version: 
`&nbsp;<ruby title="週末(しゅうまつ)"><rb>週末</rb><rt>しゅうまつ</rt></ruby>は<span class="clozewrapper">{{c1::<ruby title="何(なに)"><rb>何</rb><rt>なに</rt></ruby>かしましたか}}<!-- clozewr --></span><!-- apper -->？`

...it appropriately hides the cloze parameters, but it reads both the [ruby characters](https://en.wikipedia.org/wiki/Ruby_character) *and* the text below, so that AwesomeTTS parses it as:

    週末
    しゅうまつ
    は何
    なに
    かしましたか？

So there should be at least a parameter (if not a default config) to ignore ruby characters.

**Comment by krassowski** [comment]
Did you try to create a regular expression to remove the ruby characters? I think it should be possible with something like this:

![screenshot from 2017-10-22 16-22-07](https://user-images.githubusercontent.com/5832902/31863972-605e6cc6-b745-11e7-9739-b3a0b0a17a92.png)

**Comment by picsi** [comment]
I didn't try because don't really know regex. It'd be neat for us noobs if this was the default, though :)

**Comment by hjorthjort** [comment]
I have been using my patch in pull request #37 for months with no problem. Only missing tests. I don't really have time to add tests though. So you can get around it by cloning AwesomeTTS and patching, but that's a bit of a hassle.

**Comment by khonkhortisan** [comment]
I'm not having luck with awesometts' regex applying. Testing regex online, `<rt>.<\/rt>` should work in javascript regex, '<rt>.*<\/rt>` should work in php/pcre regex. `<rt>か</rt>` should work outside of regex for a specific character. Manually replacing `<rt>…</rt>` with `<rt><span style="display:none">[</span>…<span style="display:none">]</span></rt>` causes the right effect, but I wasn't able to do this automatically.

**Comment by lucwastiaux** [comment]
@nicolasmaia how do you generate your ruby characters ? it feels like you should have a field which doesn't have the ruby characters ? and that field could be the source for text to speech ?

**Comment by nicolasmaia** [comment]
They're generated automatically through https://ankiweb.net/shared/info/3918629684

**Comment by lucwastiaux** [comment]
OK and what field do you generate the ruby characters from ? Do you have a separate field with the original data ? Could that field then be used to generate your TTS sounds, without having to do extra processing ?

**Comment by nicolasmaia** [comment]
From the field Expression, to the field Reading. Not sure about that last question, though.

**Comment by lucwastiaux** [comment]
Why don't you run the TTS on the field Expression, instead of the field Reading ? Wouldn't this just work ?

**Comment by lucwastiaux** [comment]
closed for inactivity, feel free to re-open if necessary

**Comment by olslash** [comment]
I'd like to bump this issue since it's affecting me as well, and probably many other Japanese learners. To clarify your earlier question, @lucwastiaux, it might help to think of the expression and its reading as one and the same thing. 

Furigana is a feature of the Japanese writing system which combines characters and their readings into a single line of text, with the readings smaller and overhead. [Browsers support this using the `<ruby>`, `<rb>` and `<rt>` tags](https://www.w3.org/International/articles/ruby/markup.en.html).

Here's an example:
<h2>
これは<ruby><rb>日</rb><rt>に</rt><rb>本</rb><rt>ほん</rt><rb>語</rb><rt>ご</rt></ruby>です。
</h2>

```html
これは<ruby><rb>日</rb><rt>に</rt><rb>本</rb><rt>ほん</rt><rb>語</rb><rt>ご</rt></ruby>です。
```

You can probably see why it's not useful to have two fields here:
- the upper and lower portions are always meant to be seen together.
- the `<rt>` readings apply to specific pieces of the `<rb>` text, and are only useful in that position

So on one of my notes, I would have a field with this markup in it (generated by another plugin), that looks like this:
![image](https://user-images.githubusercontent.com/693493/113808900-dd70cb00-971b-11eb-9535-fc9a4e0f7b55.png)
and what we want is for the TTS reader to strip out the reading portion of the ruby group (the content of the `<rt>` tag) to avoid voicing duplicate sounds.

The reason this isn't easy to do today is because user-defined regex text substitution rules run _after AwesomeTTS has already stripped all html tags from the field_. Because of that, we have no way to identify and remove only readings.

**Comment by olslash** [comment]
one thing i'll note, there is a workaround: you can just have the text un-annotated, and then a second time annotated but in parens, like this:

![image](https://user-images.githubusercontent.com/693493/113809895-f24e5e00-971d-11eb-8e2f-df54c7c28067.png)

but you can see that it's not ideal, especially for long sentences

**Comment by lucwastiaux** [comment]
Thanks for explaining this in a detailed way, I agree AwesomeTTS should have a way of handling this given it's likely a lot of japanese learners will have this type of setup. I've also seen some fields containing Furigana in brackets / [ ] . Is this common as well ?
Could I ask someone to upload a deck containing Furigana using Ruby characters ? From the Anki main screen, click File → Export, click Export... . Then locate apkg file you just created, and upload it to https://www.dropbox.com/request/X9L4mhPqEcfvKFuSm2nK

**Comment by olslash** [comment]
yes, the plugin I use will also generate bracketed readings, which look like this:

![image](https://user-images.githubusercontent.com/693493/113810063-4d805080-971e-11eb-94d7-bb9abe5feefa.png)

but it's way harder to read them :)

**Comment by lucwastiaux** [comment]
A couple more questions:
- which one is the addon which generates the Furigana ?
- do you use batch generation, on-the-fly TTS or both ?
- in how much of a hurry are you to get this working ? I'm in the middle of planning an AwesomeTTS rewrite which will allow me to implement such features much more easily.

**Comment by olslash** [comment]
uploaded an example deck. 

- addon: https://ankiweb.net/shared/info/1444055400
- both, my example shows on-the-fly but I've been considering switching my decks to batch gen.
- no rush, since a workaround exists - this just smooths the UX for us a bit

thanks for the prompt help

**Comment by lucwastiaux** [comment]
thanks a lot. regarding batch vs. on the fly, just be aware that if AwesomeTTS does some replacement, that same logic won't extend to AnkiMobile or AnkiDroid (anyway AnkiDroid  is still working on supporting Anki 2.1.20's TTS tag). If your requirement is faithful reproduction of audio on mobile, you  may need to use batch generation.

**Comment by lucwastiaux** [comment]
@olslash thanks for example deck. I started looking into this. One issue i'm hitting is that Anki's TTS framework is passing HTML-stripped data to awesomeTTS so the Ruby tags are lost. I started a thread here https://forums.ankiweb.net/t/getting-raw-field-data-in-custom-tts-implementation/9603 and we'll see how it goes.

However it seems the solution is working for batch generation. @olslash is proper handling of ruby/furigana for batch generation something that's interesting to you , or you absolutely need it for on-the-fly TTS ?

**Comment by lucwastiaux** [comment]
the answer from the anki developer is: "Not sure it makes that much sense to read out both expression and reading at the same time - wouldn’t it better better if the user just used tts:kana:somefield?"

**Comment by olslash** [comment]
nice! If batch gen works correctly now, then that's great news and gives us a solution for the immediate problem. That said I'm happy to help push towards a general fix for this, since full Japanese support would be even better.

I think @dae is missing some context and assuming you want to read two different TTS utterances from a single field. Would probably help to link to my [comment ](https://github.com/AwesomeTTS/awesometts-anki-addon/issues/9#issuecomment-814585342) up above.

it does seem like it'd be trivial for them to give you a second field alongside `field_text`, like `field_text_raw`.

**Comment by lucwastiaux** [comment]
I've released awesomeTTS 1.39 which supports proper ruby handling, enabled by default. please give this a try and let me know whether that part at least works as expected.

I still want to clarify this question: you generated the Furigana data from somewhere correct ? did you start with the expression `東京` and then ran an add-on which generated `<ruby title="東京(とうきょう)"><rb>東京</rb><rt>とうきょう</rt></ruby>&nbsp;<br>` ?
If that was indeed your workflow, why not preserve the original `東京` in a field which could then be fed to the TTS engine unaltered ?

**Comment by olslash** [comment]
Just tested the batch gen changes and they sound great to me. Makes a big difference, so thanks a lot.

Yeah, you can definitely have an unrendered field without furigana on it and source the TTS tag from that, it's a decent workaround and not too hard to achieve with a bit of copy/paste. Of course, cards often have many fields with TTS readings, so if I have a card with 5 spoken fragments (perhaps a sentence and a few conjugations or vocab words in isolation), I would need 5+5 fields to use the workaround.
 
You've done a ton already to make support better for batch gen, so totally understand if you want to close this and move on with other stuff.

That said I'm still happy to help push for a fix to both fronts, which would require the ability for extension authors to get at unstripped field text and interpret it as they see fit. Seems like the Anki team really needs to do very little on their end, and then you'd be be able to support the ruby standard in a way that's unsurprising to users.

Thanks again for the effort so far.

**Comment by lucwastiaux** [comment]
I'll keep on exploring ways to support this for on-the-fly. I migrated AwesomeTTS to use the Anki 2.1.20 tts tag standard, but it seems this is one of the limitations with this approach. Previously AwesomeTTS could access the raw field data.
I'm planning a rewrite of AwesomeTTS and i'll try to plan for this sort of feature.
